Added options menu

andrewhong04 opened the pull request:
I need some constructive criticism on my code lol.

andrewhong04 committed:
Added options state

This took a little too long.. lol

andrewhong04 committed:
Added a 'must restart" message

andrewhong04 committed:
Merge branch 'Tuxemon:development' into development

bitcraft commented on tuxemon/states/control/__init__.py:
```diff
@@ -0,0 +1,129 @@
+            item = MenuItem(image, label, None, None)
+            self.add(item)
+
+    def process_event(self, event: PlayerInput) -> Optional[PlayerInput]:
```
please add docstrings like this for this and the other new functions/methods added in the pr
```
"""
Some short description

Parameters:
    event: Player Input to be checked

"""
```

vnmabus replied:
This is an inherited method. The docstrings are automatically inherited, so I would only override them if there is something to change/add.

bitcraft replied:
good point, thank you

bitcraft commented on tuxemon/states/control/__init__.py:
```diff
@@ -0,0 +1,129 @@
+"""This module contains the Options state
+"""
+from __future__ import annotations
+\
```
random `\`

bitcraft commented on mods/tuxemon/db/locale/en_US.json:
```diff
@@ -80,6 +80,13 @@
+    "menu_left_key": "Left Key",
+    "menu_right_key": "Right Key",
+    "menu_down_key": "Down Key",
+    "menu_primary_select_key": "P-Select Key",
```
I've been using "Action" and "Cancel" for the button names.  Does that seem ok?  I'm not sure is "P-Select" and "S-Select" are clear enough.

andrewhong04 replied:
Oh, I didn't know the names that were used in the actual code. I'll rename them now.

bitcraft commented on tuxemon/config.py:
```diff
@@ -107,6 +107,15 @@ def __init__(self, config_path: Optional[str] = None) -> None:
         self.player_walkrate = cfg.getfloat("player", "player_walkrate")  # tiles/second
         self.player_runrate = cfg.getfloat("player", "player_runrate")  # tiles/second
 
+        # [controls]
```
This is duplicated by existing code in config.py
```python
        # input config (None means use default for the platform)
        self.gamepad_deadzone = 0.25
        self.gamepad_button_map = None
        self.keyboard_button_map = get_custom_pygame_keyboard_controls(cfg)
```

andrewhong04 replied:
I made a very crude solution to this issue.. Let me know if this looks a little messy after I send the commit.

bitcraft commented on tuxemon/states/control/__init__.py:
```diff
@@ -0,0 +1,129 @@
+
+    def process_event(self, event: PlayerInput) -> Optional[PlayerInput]:
+        pressed_key = None
+        if pygame.key.get_pressed()[pygame.K_UP]: pressed_key = "up"
```
please add this note
```
# must use get_pressed because the events do not contain references to pygame events
```

bitcraft commented on tuxemon/states/control/__init__.py:
```diff
@@ -0,0 +1,129 @@
+    config.back = cfg_file.get("controls", "back")
+
+class SetKeyState(PopUpMenu):
+    """This state is responsible for setting the input keys"""
```
Please add a comment in the docstring stating that this only works for pygame events.

bitcraft commented on tuxemon/states/control/__init__.py:
```diff
@@ -0,0 +1,129 @@
+
+    def startup(self, **kwargs: Any) -> None:
+        # TODO: update the menu once a control key has been changed
+        # TODO: remove the need to restart the game for changes to take place
```
To apply this without restarting, you need to update the existing input map for the keyboard.
The game client has an input manager, which aggregates the various input methods and produces virtual input events.  So you will need to access the game client, the input manager, find the keyboard input, finally, modify the input map.  The game client can be accessed using the "local session", i've included an example of it below

https://github.com/Tuxemon/Tuxemon/blob/development/tuxemon/client.py#L93
https://github.com/Tuxemon/Tuxemon/blob/development/tuxemon/platform/events.py#L93
https://github.com/Tuxemon/Tuxemon/blob/development/tuxemon/platform/platform_pygame/events.py#L45

andrewhong04 replied:
I tried your solutions, but they don't seem to work unless I'm missing something. Any other ideas on how to fix it? I've sent a new commit.

andrewhong04 replied:
My bad, I missed something.

bitcraft commented on tuxemon/states/control/__init__.py:
```diff
@@ -0,0 +1,129 @@
+cfg_file = config.TuxemonConfig(paths.USER_CONFIG_PATH).cfg
+config = prepare.CONFIG
+
+def reload_control_config():
```
keyboard control map loading is already here
https://github.com/Tuxemon/Tuxemon/blob/development/tuxemon/config.py#L131

bitcraft reviewed (commented):
Looks great so far.  Very nice use of the menus.  The workarounds for getting the keys are fine with me.

andrewhong04 committed:
optimal changes

Renamed P-Select & S-Select to Action and Cancel
Removed random \

vnmabus commented on tuxemon/states/control/__init__.py:
```diff
@@ -0,0 +1,129 @@
+            if isinstance(pressed_key, int): pressed_key_str = pygame.key.name(pressed_key)
+
+            cfg_file.set("controls", self.input, pressed_key_str)
+            with open(paths.USER_CONFIG_PATH, "w") as fp:
```
Do we want to rewrite the file every time?

novialriptide replied:
I'm not sure how else to implement it.

andrewhong04 replied:
I'm not sure how else to implement it

vnmabus replied:
I think in most programs the control remapping only takes place after you leave the control remapping config window. Thus, the configuration is only written once, and if you select a wrong key by mistake you can still change it with the old controls.

andrewhong04 replied:
Oh, I'll try that.

andrewhong04 replied:
Fixed!

andrewhong04 committed:
added docstrings. not sure what to do from here.

andrewhong04 committed:
Merge branch 'development' of https://github.com/novialriptide/Tuxemon into development

andrewhong04 committed:
Merge branch 'Tuxemon:development' into development

andrewhong04 committed:
removed duplicate controls and cleaned up some code

andrewhong04 committed:
Merge branch 'development' of https://github.com/novialriptide/Tuxemon into development

andrewhong04 committed:
Fixed some bugs

bitcraft commented on tuxemon/client.py:
```diff
@@ -93,18 +93,13 @@ def __init__(self, config: TuxemonConfig) -> None:
-            config.gamepad_button_map,
-            config.gamepad_deadzone,
-        )
+        gamepad = PygameGamepadInput(config.gamepad_button_map, config.gamepad_deadzone)
```
try to avoid formatting changes if possible, but you don't *have* to revert this.

bitcraft commented on tuxemon/config.py:
```diff
@@ -128,17 +128,15 @@ def __init__(self, config_path: Optional[str] = None) -> None:
         button_value: Optional[int] = getattr(buttons, key, None)
         event_value: Optional[int] = getattr(events, key, None)
-        for each in values.split(", "):
+        for each in values.split(", "): # used incase of multiple keys assigned to 1 method
```
try to avoid long inline comments.  instead, put the over the line they are describing.

bitcraft commented on tuxemon/states/control/__init__.py:
```diff
@@ -0,0 +1,158 @@
+        
+        if isinstance(pressed_key, str): pressed_key_str = pressed_key
+        if isinstance(pressed_key, int): pressed_key_str = pygame.key.name(pressed_key)
+        print(invalid_keys, pressed_key_str)
```
please make sure debug prints are removed

bitcraft commented on tuxemon/config.py:
```diff
@@ -128,17 +128,15 @@ def __init__(self, config_path: Optional[str] = None) -> None:
-def get_custom_pygame_keyboard_controls(
-    cfg: configparser.ConfigParser,
-) -> Mapping[Optional[int], int]:
+def get_custom_pygame_keyboard_controls(cfg: configparser.ConfigParser) -> Mapping[Optional[int], int]:
```
try to avoid changing the formatting.  lines should not exceed 80 characters

andrewhong04 committed:
Merge branch 'Tuxemon:development' into development

andrewhong04 committed:
Removed the need to restart the game to update controls

andrewhong04 committed:
Merge branch 'Tuxemon:development' into development

andrewhong04 committed:
added menu refresh after changes

andrewhong04 committed:
fixed glitch where arrow keys would crash the game

andrewhong04 commented:
Hi, I've implemented the controls change menu, but it uses "|" to create columns. Is there a cleaner way of doing this?

bitcraft commented on tuxemon/menu/menu.py:
```diff
@@ -305,9 +312,7 @@ def reload_sounds(self) -> None:
-    def shadow_text(
-        self,
-        text: str,
+    def shadow_text(self, text: str,
```
Please do not change formatting

bitcraft commented on tuxemon/platform/events.py:
```diff
@@ -86,10 +86,7 @@ class InputHandler(ABC, Generic[_InputEventType]):
-        self,
-        event_map: Optional[Mapping[Optional[int], int]] = None,
-    ) -> None:
+    def __init__(self, event_map: Optional[Mapping[Optional[int], int]] = None) -> None:
```
is your editor changing these?

bitcraft committed:
Update CONTRIBUTING.md

bitcraft committed:
Update CONTRIBUTING.md

bitcraft committed:
Update fmt.sh

pep8 formatting

andrewhong04 committed:
formatting changes reverted

more format changes

more format reverts

more reverts

more reverts

reverted docstring newlines

added newline at the end of state.py

more reverts

more reverts

reverted docstring newlines

added newline at the end of state.py

added docstrings + pep8 formatting

andrewhong04 committed:
formatting changes reverted

andrewhong04 commented:
I reverted most of the format changes I've made to the files, that was my bad.

bitcraft commented on tuxemon/states/control/__init__.py:
```diff
@@ -0,0 +1,165 @@
+        if event.button == buttons.BACK:
+            self.reload_controls()
+
+        return super().process_event(event)
```
pep8: newline needed

bitcraft commented on tuxemon/states/start/__init__.py:
```diff
@@ -81,12 +81,12 @@ def startup(self, **kwargs: Any) -> None:
         def change_state(
             state: Union[State, str],
-            **change_state_kwargs: Any,
+            **change_state_kwargs: Any
```
commas needed (see original)

bitcraft commented on tuxemon/states/start/__init__.py:
```diff
@@ -81,12 +81,12 @@ def startup(self, **kwargs: Any) -> None:
                 self.client.push_state,
                 state,
-                **change_state_kwargs,
+                **change_state_kwargs
```
commas needed (see original)

bitcraft commented on tuxemon/platform/platform_pygame/events.py:
```diff
@@ -35,6 +35,14 @@ def add_input(self, player: int, handler: InputHandler[Any]) -> None:
         """
         self._inputs[player].append(handler)
 
+    def set_input(
```
is this overloading the super class method?  looks unchanged

bitcraft commented on tuxemon/platform/events.py:
```diff
@@ -58,6 +58,14 @@ def release_controls(self) -> Generator[PlayerInput, None, None]:
             for inp in value:
                 yield from inp.virtual_stop_events()
 
+    def set_input(
```
docstring needed

andrewhong04 committed:
Merge branch 'Tuxemon:development' into development

bitcraft reviewed (approved):
looks good, thank you for working with us to make all those changes!
